### ISSUE TITLE
Unused workspaces in PipelineRun

### ISSUE THREAD

**Issue by jimmyjones2** [tektoncd/pipeline#6856]
If I create a PipelineRun containing a workspace that is unused by the Pipeline, there are no errors and it completes successfully. I can't find this behaviour documented.

I've got a fairly generic process that creates PipelineRuns, but depending on the Pipeline chosen, some of the workspaces may be unused. It is safe to rely on this behaviour?

Example:
```
apiVersion: tekton.dev/v1beta1
kind: Pipeline
metadata:
  name: clone-read
  namespace: default
spec:
  description: |
    This pipeline clones a git repo, then echoes the README file to the stout.
  params:
  - description: The git repo URL to clone from.
    name: repo-url
    type: string
  tasks:
  - name: fetch-source
    params:
    - name: url
      value: $(params.repo-url)
    taskRef:
      kind: Task
      name: git-clone
    workspaces:
    - name: output
      workspace: shared-data
  - name: show-readme
    runAfter:
    - fetch-source
    taskRef:
      kind: Task
      name: show-readme
    workspaces:
    - name: source
      workspace: shared-data
  workspaces:
  - description: |
      This workspace contains the cloned repo files, so they can be read by the
      next task.
    name: shared-data
---
apiVersion: tekton.dev/v1beta1
kind: PipelineRun
metadata:
  generateName: clone-read-run-
  namespace: default
spec:
  params:
  - name: repo-url
    value: https://github.com/tektoncd/website.git
  pipelineRef:
    name: clone-read
  podTemplate:
    securityContext:
      fsGroup: 65532
  serviceAccountName: default
  timeout: 1h0m0s
  workspaces:
  - name: foo
    configMap:
      name: foo
  - name: shared-data
    volumeClaimTemplate:
      metadata:
        creationTimestamp: null
      spec:
        accessModes:
        - ReadWriteOnce
        resources:
          requests:
            storage: 1Gi
      status: {}
```

**Comment by pritidesai** [comment]
There is no such validation which restricts adding unused workspaces in `pipelineRun.spec.workspaces`. We validate workspaces for `isExist()` i.e. a workspace defined in a pipeline must exist in pipelineRun. And there is a validation of `Unique()` among the workspaces defined in  `pipelineRun.spec.workspaces`.

isExist():

https://github.com/tektoncd/pipeline/blob/7b23f931d4f93aab8688f87536a75b50085f1285/pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go#L450-L465

Unique:

https://github.com/tektoncd/pipeline/blob/7b23f931d4f93aab8688f87536a75b50085f1285/pkg/apis/pipeline/v1/pipelinerun_validation.go#L98-L107

It might be accidental to allow extra workspaces but I do not foresee restricting such specification.

@tektoncd/core-collaborators shall we document this behavior and allow specifying any number of workspaces in `pipelineRun.spec.workspaces`?

**Comment by lbernick** [comment]
This seems like a bug to me. I'm not sure there's a good reason to support specifying extra workspaces in the pipelinerun, and I would expect most pipelinerun authors to prefer fail-fast behavior, i.e. rejecting pipelineruns with extra workspaces since they're most likely misconfigured.

**Comment by jerop** [comment]
It seems supporting extra `Workspaces` was intentional for the same reason that we added support for extra `Parameters` - https://github.com/tektoncd/pipeline/issues/2513: 

> Imagine you have a system creating PipelineRuns automatically, and each PipelineRun uses a different Pipeline which has a different set of parameters. The system has a number of parameters it knows how to provide, e.g. the service account and url of a bucket the Pipelines can push to (but not all Pipelines need to push to a bucket).
> 
> It would be very convenient if the system creating the PipelineRuns could provide all the parameters it has available to all PipelineRuns, and the Pipelines could use the ones they need and ignore the rest.

While I see the value of fail-fast as @lbernick suggested, I think we should document this behavior as some users could be depending on it. Here is the documentation for extra `Parameters` for reference: https://github.com/tektoncd/pipeline/blob/21828f35ad3e31da5c0eb8b778cb8166d5e02fcb/docs/pipelineruns.md?plain=1#L269-L272

**Comment by jimmyjones2** [comment]
Fixed in #6917
